### ISSUE TITLE
cache lock is not released on ctrl-c

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#1620]
```
$ /usr/local/bin/borg create --compression lz4 -v --stats --one-file-system --show-rc --exclude-from excludes.txt '/mnt/repo::{hostname}-{user}-{now:%Y-%m-%d_%H:%M:%S}' /root /etc /mnt/data/homes/m /mnt/data/homes/t
^CKeyboard interrupt.
Traceback (most recent call last):
  File "borg/archiver.py", line 249, in do_create
  File "borg/archiver.py", line 223, in create_inner
  File "borg/archiver.py", line 303, in _process
  File "borg/archiver.py", line 303, in _process
  File "borg/archiver.py", line 303, in _process
  File "borg/archiver.py", line 303, in _process
  File "borg/archiver.py", line 303, in _process
  File "borg/archiver.py", line 303, in _process
  File "borg/archiver.py", line 277, in _process
  File "borg/archive.py", line 655, in process_file
  File "borg/cache.py", line 416, in file_known_and_unchanged
  File "msgpack/__init__.py", line 47, in packb

KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
 File "borg/archiver.py", line 1684, in main
 File "borg/archiver.py", line 1621, in run
 File "borg/archiver.py", line 85, in wrapper
 File "borg/archiver.py", line 249, in do_create

KeyboardInterrupt
        
Platform: Linux v 3.13.0-92-generic #139-Ubuntu SMP Tue Jun 28 20:42:32 UTC 2016 i686 i686
Linux: debian jessie/sid
Borg: 1.0.7  Python: CPython 3.5.2
sys.argv: ['/usr/local/bin/borg', 'create', '--compression', 'lz4', '-v', '--stats', '--one-file-system', '--show-rc', '--exclude-from', 'excludes.txt', '/mnt/repo::{hostname}-{user}-{now:%Y-%m-%d_%H:%M:%S}', '/root', '/etc', '/mnt/data/homes/m', '/mnt/data/homes/t']
SSH_ORIGINAL_COMMAND: None

terminating with error status, rc 2

$ /usr/local/bin/borg prune --prefix '{hostname}-{user}-' --stats --list -v --keep-with 1d --keep-daily 7 --keep-weekly 4 --keep-monthly 12 --keep-yearly 1 /mnt/repo
Failed to create/acquire the lock /mnt/.../cache/2c...35/lock.exclusive (timeout).
```

**Comment by ThomasWaldmann** [comment]
Hmm, is it because there are two KeyboardInterrupts?

**Comment by enkore** [comment]
Hmm, the second KeyboardInterrupt is exactly on the last line of the `with Cache(...)` block. That seems a bit odd, given that `__exit__` immediately releases any lock.

**Comment by capi** [comment]
The original report came from me via IRC. It is very likely that I've hit `Ctrl-C` twice.

**Comment by ThomasWaldmann** [comment]
Maybe we should disable any further Ctrl-C in a signal handler?
And (if we do not exit anyway), re-enable it at the end of the KeyboardInterrupt exception handler?

**Comment by ThomasWaldmann** [comment]
Code:
```
from time import sleep

try:
    sleep(10)
except KeyboardInterrupt:
    print("kbi handler start")
    sleep(10)
    print("kbi handler end")
```
Behaviour when quickly pressing Ctrl-C multiple times:
```
$ python3.4 ctrlc.py 
^Ckbi handler start
^CTraceback (most recent call last):
  File "ctrlc.py", line 4, in <module>
    sleep(10)
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "ctrlc.py", line 7, in <module>
    sleep(10)
KeyboardInterrupt
```
So, this obviously does not reach the end of the KBI exception handler block because another KBI interrupts it again...

**Comment by ThomasWaldmann** [comment]
This works better:
```
import signal
import time

def sigint(sig, frame):
    signal.signal(signal.SIGINT, signal.SIG_IGN)
    raise KeyboardInterrupt

signal.signal(signal.SIGINT, sigint)

try:
    time.sleep(5)
except KeyboardInterrupt:
    print("kbi handler start")
    time.sleep(5)
    print("kbi handler end")
```

**Comment by ThomasWaldmann** [comment]
This issue is not limited to sigint, but happens in the same way for sigterm, sighup and maybe others.

**Comment by ThomasWaldmann** [comment]
We already have signal_handler contextmanager in master, so we just need sig_exc_handler additionally (in master, and both in 1.0-maint), plus some exception classes:
```
import contextlib
import signal
import time


class SignalException(BaseException):
    """base class for all signal-based exceptions"""


class SigInt(SignalException):
    """raised on SIGINT signal"""


class SigHup(SignalException):
    """raised on SIGHUP signal"""


class SigTerm(SignalException):
    """raised on SIGTERM signal"""


def sig_exc_handler(exc_cls):
    def handler(sig_no, frame):
        # setting SIG_IGN avoids that an incoming second signal of this
        # kind would raise a 2nd exception while we still process the
        # exception handler for exc_cls for the 1st signal.
        signal.signal(sig_no, signal.SIG_IGN)
        raise exc_cls(sig_no)

    return handler


@contextlib.contextmanager
def signal_handler(sig_no, handler):
    orig_handler = signal.signal(sig_no, handler)
    try:
        yield
    finally:
        # restore original handler / potentially un-ignore this signal
        signal.signal(sig_no, orig_handler)


with signal_handler(signal.SIGINT, sig_exc_handler(SigInt)), \
     signal_handler(signal.SIGHUP, sig_exc_handler(SigHup)), \
     signal_handler(signal.SIGTERM, sig_exc_handler(SigTerm)):
    try:
        time.sleep(50)
    except SignalException as e:
        print("sig %s handler start" % e)
        time.sleep(5)
        print("sig %s handler end" % e)

print("orig handler restored")
time.sleep(5)
print("program end")
```

**Comment by ThomasWaldmann** [comment]
any ideas for better names for signal_handler and sig_exc_handler?

**Comment by enkore** [comment]
raising_signal_handler?

**Comment by ThomasWaldmann** [comment]
ok, that is better. now it is just that both are called ...signal_handler, but the signal_handler is not really a signal_handler, it is rather a signal_handler_manager - but that sounds too much like Java.

**Comment by enkore** [comment]
The @contextmanager and use as context manager distinguishes that okay-ish from each other

**Comment by ThomasWaldmann** [comment]
fixed by #1631.
